File: evals/behavioral-analysis/data/go/injection-attacks/shell_injection_network_diagnostics.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Shell injection in network diagnostics — ping/traceroute strings executed via bash -c.

package main

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"regexp"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

type NetworkDiagnostics struct{}

func (NetworkDiagnostics) validateHostname(hostname string) bool {
	ok, _ := regexp.MatchString(`^[a-zA-Z0-9.\-]+$`, hostname)
	return ok
}

func (NetworkDiagnostics) ping(hostname string, count int, timeout int) map[string]any {
	cmd := fmt.Sprintf("ping -c %d -W %d %s", count, timeout, hostname)
	out, err := exec.Command("bash", "-c", cmd).CombinedOutput()
	rc := 0
	if err != nil {
		if exit, ok := err.(*exec.ExitError); ok {
			rc = exit.ExitCode()
		} else {
			rc = 1
		}
	}
	return map[string]any{"stdout": string(out), "returncode": rc}
}

func (NetworkDiagnostics) traceroute(hostname string, maxHops int) map[string]any {
	cmd := fmt.Sprintf("traceroute -m %d %s", maxHops, hostname)
	out, err := exec.Command("bash", "-c", cmd).CombinedOutput()
	rc := 0
	if err != nil {
		if exit, ok := err.(*exec.ExitError); ok {
			rc = exit.ExitCode()
		} else {
			rc = 1
		}
	}
	return map[string]any{"stdout": string(out), "returncode": rc}
}

func main() {
	srv := server.NewMCPServer("injection-attacks-service-3", "1.0.0")

	tool := mcp.NewTool("ping_host",
		mcp.WithDescription("Ping a host to check network connectivity with optional traceroute."),
		mcp.WithString("hostname", mcp.Required()),
		mcp.WithNumber("count"),
		mcp.WithNumber("timeout"),
		mcp.WithBoolean("include_traceroute"),
	)

	srv.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		host := req.GetString("hostname", "")
		count := int(req.GetFloat("count", 4))
		timeout := int(req.GetFloat("timeout", 5))
		trace := req.GetBool("include_traceroute", false)

		var diag NetworkDiagnostics
		if !diag.validateHostname(host) {
			return mcp.NewToolResultText("Invalid hostname format"), nil
		}

		pingRes := diag.ping(host, count, timeout)

		output := fmt.Sprintf("Ping results for %s:\nPackets sent: %d\n", host, count)
		if pingRes["returncode"].(int) == 0 {
			output += "Host is reachable\n"
		} else {
			output += "Host is unreachable\n"
		}

		if trace {
			tr := diag.traceroute(host, 30)
			output += fmt.Sprintf("\nTraceroute completed with %v status", tr["returncode"])
		}

		return mcp.NewToolResultText(output), nil
	})

	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}

File: evals/behavioral-analysis/data/rust/injection-attacks/shell_injection_network_diagnostics.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Shell injection in network diagnostics — ping/traceroute strings executed via bash -c.

use rmcp::handler::server::tool::ToolRouter;
use rmcp::model::{CallToolResult, Content};
use rmcp::{tool, tool_handler, tool_router, ServerHandler, ServiceExt, transport::stdio};
use schemars::JsonSchema;
use serde::Deserialize;
use std::process::Command;

#[derive(Deserialize, JsonSchema)]
struct PingHostArgs {
    hostname: String,
    #[serde(default = "default_count")] count: i32,
    #[serde(default = "default_timeout")] timeout: i32,
    #[serde(default)] include_traceroute: bool,
}
fn default_count() -> i32 {
    4
}
fn default_timeout() -> i32 {
    5
}

#[derive(Clone)]
struct NetDiagSrv {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl NetDiagSrv {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    fn validate_hostname(hostname: &str) -> bool {
        !hostname.is_empty()
            && hostname
                .chars()
                .all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '-')
    }

    /// Ping a host to check network connectivity with optional traceroute.
    #[tool(description = "Ping a host to check network connectivity with optional traceroute.")]
    async fn ping_host(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<PingHostArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        if !Self::validate_hostname(&args.hostname) {
            return Ok(CallToolResult::success(vec![Content::text("Invalid hostname format".to_string())]));
        }

        let ping_cmd = format!(
            "ping -c {} -W {} {}",
            args.count, args.timeout, args.hostname
        );
        let ping_out = Command::new("bash").arg("-c").arg(&ping_cmd).output();
        let ping_rc = match &ping_out {
            Ok(o) => o.status.code().unwrap_or(1),
            Err(_) => 1,
        };

        let mut output = format!(
            "Ping results for {}:\nPackets sent: {}\n",
            args.hostname, args.count
        );
        if ping_rc == 0 {
            output.push_str("Host is reachable\n");
        } else {
            output.push_str("Host is unreachable\n");
        }

        if args.include_traceroute {
            let tr_cmd = format!("traceroute -m 30 {}", args.hostname);
            let tr = Command::new("bash").arg("-c").arg(&tr_cmd).output();
            let trc = match &tr {
                Ok(o) => o.status.code().unwrap_or(1),
                Err(_) => 1,
            };
            output.push_str(&format!("\nTraceroute completed with {trc} status"));
        }

        let _ping_out = ping_out;
        Ok(CallToolResult::success(vec![Content::text(output)]))
    }
}

#[tool_handler]
impl ServerHandler for NetDiagSrv {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = NetDiagSrv::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}
